From 7b270397adfb0d65b5f0d9830b94696296970ae4 Mon Sep 17 00:00:00 2001
From: Sibin Silvester <sibinsilva@gmail.com>
Date: Sun, 12 Jul 2026 15:09:49 +0000
Subject: [PATCH] rkpd: skip IRPC registration when HAL is undeclared

`RegistrationProxy currently attempts to create an rkpd registration even when
android.hardware.security.keymint.IRemotelyProvisionedComponent/default is not
declared by the device VINTF. On devices without this HAL, the registration
path fails early and can contribute to boot-time instability. Add a declaration
check and return a controlled error when the IRPC HAL is absent, allowing the
system to continue booting without depending on vendor support that does not
exist.`


Change-Id: I2199c873f643f7184d2f7540cb5d63f9357fa1ce
---
 .../security/rkp/service/RegistrationProxy.java  | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/system-server/src/android/security/rkp/service/RegistrationProxy.java b/system-server/src/android/security/rkp/service/RegistrationProxy.java
index 7385baf..408b7f6 100644
--- a/system-server/src/android/security/rkp/service/RegistrationProxy.java
+++ b/system-server/src/android/security/rkp/service/RegistrationProxy.java
@@ -36,6 +36,7 @@ import android.os.IBinder;
 import android.os.OperationCanceledException;
 import android.os.OutcomeReceiver;
 import android.os.RemoteException;
+import android.os.ServiceManager;
 import android.os.UserHandle;
 import android.util.Log;
 
@@ -67,6 +68,15 @@ public class RegistrationProxy {
     static final String TAG = "RegistrationProxy";
     IRegistration mBinder;
 
+    private static boolean isIrpcDeclared(@NonNull String irpcName) {
+        try {
+            return ServiceManager.isDeclared(irpcName);
+        } catch (Throwable t) {
+            Log.w(TAG, "Failed to query ServiceManager for " + irpcName, t);
+            return false;
+        }
+    }
+
     /** Deals with the {@code ServiceConnection} lifetime for the rkpd bound service. */
     private static class IRemoteProvisioningConnection implements ServiceConnection {
         @GuardedBy("this")
@@ -180,6 +190,12 @@ public class RegistrationProxy {
             @NonNull String irpcName, @NonNull Duration bindTimeout,
             @NonNull @CallbackExecutor Executor executor,
             @NonNull OutcomeReceiver<RegistrationProxy, Exception> receiver) {
+        if (!isIrpcDeclared(irpcName)) {
+            Log.w(TAG, "IRPC HAL not declared in VINTF, skip registration for " + irpcName);
+            executor.execute(() -> receiver.onError(
+                    new RemoteException("IRPC not declared: " + irpcName)));
+            return;
+        }
         try {
             // The connection object is used to get exactly one IRegistration binder. Once we
             // get it, we unbind the connection. This allows the bound service to be terminated
-- 
2.43.0

